src/mavsdk/core/mavlink_parameter_receiver.cpp

```diff
@@ -106,176 +118,177 @@ std::map<std::string, ParamValue> MavlinkParameterReceiver::retrieve_all_server_
-        _work_queue.push_back(new_work);
-        std::lock_guard<std::mutex> lock(_param_changed_subscriptions_mutex);
-
-        for (const auto& subscription : _param_changed_subscriptions) {
-            if (subscription.param_name != safe_param_id) {
-                continue;
-            }
-
-            if (!subscription.any_type &&
-                !subscription.value_type.is_same_type(new_work->param_value)) {
-                LogErr() << "Received wrong param type in subscription for "
-                         << subscription.param_name;
-                continue;
-            }
-
-            call_param_changed_callback(subscription.callback, new_work->param_value);
-        }
-    } else {
-        LogWarn() << "Invalid Param Ext Set ID Request: " << safe_param_id;
+    const std::string safe_param_id = extract_safe_param_id(set_request.param_id);
+    if(safe_param_id.empty()){
+        // TODO support int as index, as done in the mavlink specification
+        LogWarn() << "Invalid Param Set ID Request: " << safe_param_id;
+        return;
+    }
+    // Use the ID
+    ParamValue value;
+    if (!value.set_from_mavlink_param_ext_set(set_request)) {
+        LogWarn() << "Invalid Param Ext Set Request: " << safe_param_id;
+        return;
     }
+    process_param_set_internally(safe_param_id,value, true);
 }
+
 void MavlinkParameterReceiver::process_param_request_read(const mavlink_message_t& message)
 {
     mavlink_param_request_read_t read_request{};
     mavlink_msg_param_request_read_decode(&message, &read_request);
 
-    std::string param_id = extract_safe_param_id(read_request.param_id);
-
     if (read_request.param_index == -1) {
-        auto safe_param_id = extract_safe_param_id(read_request.param_id);
+        const auto safe_param_id= extract_safe_param_id(read_request.param_id);
         LogDebug() << "Request Param " << safe_param_id;
-
         // Use the ID
         if (_all_params.find(safe_param_id) != _all_params.end()) {
-            auto new_work = std::make_shared<WorkItem>(_timeout_s_callback());
-            new_work->type = WorkItem::Type::Value;
-            new_work->param_name = safe_param_id;
+            // Make sure we are not forwarding an extended param via the "normal" param messages.
+            const auto param_value= _all_params.at(safe_param_id);
+            if(param_value.needs_extended()){
+                LogDebug()<<"Not forwarding param"<<safe_param_id<<" since it needs extended";
+                return;
+            }
+            auto new_work = std::make_shared<WorkItem>(WorkItem::Type::Value,safe_param_id,false);
             new_work->param_value = _all_params.at(safe_param_id);
-            new_work->extended = false;
             _work_queue.push_back(new_work);
         } else {
             LogDebug() << "Missing Param " << safe_param_id;
         }
+    }else{
+        // TODO the server should be able to handle requests with a int index instead of string as param_id.
```

Comment:
:thinking:

Reply:
Actually, to show why this becomes so complicated with the current support for extended and non-extended parameters - we'd need.
```
struct Parameter{
        // unique parameter id
        const std::string param_id;
        // unique parameter index when looked at from a non-extended param perspective
        const int param_idx_non_extended;
        // unique parameter index when looked at from a extended param perspective
        const int param_idx_extended;
        ParamValue param_value;
    };
```

Reply:
I think we need to lists. One normal one and one for the extended ones.

Reply:
Well, that is also not ideal, since then on each set / get operation we'd have to check 2 lists, comlicating code even further. R.n I am experimenting on the server side, if using a different type of data layout would make more sense:
`std::vector<ParamValue> _all_params`
and 
```
std::map<std::string,uint32_t> _string_id_to_index_non_extended.
std::map<std::string,uint32_t> _string_id_to_index_extended.
```

The reasoning behind that: We need 2 types of indexing anyways, if we want the param_server to be fully compatible with everything the mavlink protocoll says - by unique index or unique string id. But I am not quite sure yet what creates the most simple code in this regard.